src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -648,22 +648,31 @@ public bool IsFrozen
         internal bool IsAnyUpstreamFrozen()
         {
             bool ret = false;
-            return CheckIfAnyUpstreamNodeIsFrozen(this, ref ret);
+            List<NodeModel> nodes = new List<NodeModel>();
+            return CheckIfAnyUpstreamNodeIsFrozen(this, nodes, ref ret);
         }
 
-        private bool CheckIfAnyUpstreamNodeIsFrozen(NodeModel node, ref bool ret)
+        private bool CheckIfAnyUpstreamNodeIsFrozen(NodeModel node, List<NodeModel> nodes, ref bool ret)
         {             
             var sets = node.InputNodes.Values;
             var inpNodes = sets.Where(x => x != null).Select(z => z.Item2).Distinct();
             foreach (var inode in inpNodes)
             {
+                //If there is a cyclic-dependency, stop traversing this branch
+                if (nodes.Contains(inode))
+                {
+                    continue;
+                }
+
                 if (inode.isFrozenExplicitly)
                 {
                     ret = true;
                     break;
                 }
 
-                CheckIfAnyUpstreamNodeIsFrozen(inode, ref ret);
+                List<NodeModel> newNodes = new List<NodeModel>(nodes);
```

Comment:
If there are multiple upstream branches, and if a node in one branch is added to this list, it will be added to another list as well. Assuming an upstream node provides input for more than one branches, although it is not a cyclic graph, it will be taken as one.